app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt

```diff
@@ -0,0 +1,209 @@
+                imageWidth = image.height
+                imageHeight = image.width
+            }
+            // If the image state has changed, update the barcode overlay
+            if (_barcodeStateState.value.sourceImageWidth != imageWidth ||
+                _barcodeStateState.value.sourceImageHeight != imageHeight ||
+                _barcodeStateState.value.isFlipped != isImageFlipped
+            ) {
+                _barcodeStateState.value = _barcodeStateState.value.copy(
+                    sourceImageWidth = imageWidth,
+                    sourceImageHeight = imageHeight,
+                    isFlipped = isImageFlipped
+                )
+            }
+            // Pass the image to the barcode scanner then assign a completion listener that will
+            // close the image proxy to free-up the analysis pipeline
+            barcodeScanner.process(InputImage.fromMediaImage(it, image.imageInfo.rotationDegrees))
+                .addOnSuccessListener(TaskExecutors.MAIN_THREAD, this)
+                .addOnFailureListener(this)
+                .addOnCompleteListener {
+                    image.close()
+                }
+        }
+    }
+
+    override fun onSuccess(barcodes: List<Barcode>) {
+        // Support for a single barcode is artificial and makes displaying overlay data easier
+        val newBarcode = barcodes.firstOrNull()
+        // We wanna apply some heuristics to improve the UX in case the operator has a shacking
+        // handle and the barcode detector can make a accurate read every single frame
+        _barcodeStateState.value = if (currentBarcode == null && newBarcode != null) {
+            _barcodeStateState.value.copy(barcode = newBarcode)
+        } else if (newBarcode?.rawValue != currentBarcode?.rawValue) {
+            val now = SystemClock.elapsedRealtime()
+            if (_elapsedMillis == Long.MAX_VALUE) {
+                _elapsedMillis = now
+            }
+            if ((now - _elapsedMillis) >= MAX_BARCODE_DWELL_MS) {
+                _elapsedMillis = Long.MAX_VALUE
+                _barcodeStateState.value.copy(barcode = newBarcode)
+            } else {
+                // Keep returning the current barcode
+                _barcodeStateState.value.copy(barcode = newBarcode)
+            }
+        } else {
+            _elapsedMillis = Long.MAX_VALUE
+            // Keep updating the state to have an accurate bounding box
+            _barcodeStateState.value.copy(barcode = newBarcode)
+        }
+    }
+
+    override fun onFailure(exception: Exception) {
+        _barcodeStateState.value = _barcodeStateState.value.copy(barcode = null)
+    }
+
+    companion object {
+        private const val MAX_BARCODE_DWELL_MS = 500
+    }
+}
+
+data class BarcodeAnalysisState(
+    val sourceImageWidth: Int = 0,
+    val sourceImageHeight: Int = 0,
+    val isFlipped: Boolean = false,
+    val barcode: Barcode? = null
+) {
+    val hasBarcodes: Boolean = barcode != null
+}
```

Comment:
Let's move data class and sealed class to the top of the file.  To be consistent in the project (similar to MainViewModel).
If it's not a part of ScannerViewModel then it should be it's own file somewhere in /data/

Reply:
I've moved the data class to top of the file to match the style in MainViewModel.